### ISSUE TITLE
router:  performance improvement around route matching: remove per-route cost of repeative work on request path/headers.

### ISSUE THREAD

**Issue by stevenzzzz** [envoyproxy/envoy#21615]
*Title*: performance improvement around route matching

*Description*:
Seems like some low hanging fruits for route matching performance improvement: 

// In bool RouteEntryImplBase::matchRoute:

```
//  we are initializing parsing QueryParams out of request path per route (if query parameters configured in the route): 
  if (!config_query_parameters_.empty()) {
    Http::Utility::QueryParams query_parameters =
        Http::Utility::parseQueryString(headers.getPathValue());
    matches &= ConfigUtility::matchQueryParams(query_parameters, config_query_parameters_);
    if (!matches) {
      return false;
    }
  }

```

In PathMatcher::match(const absl::string_view path) there is a per-route cost of stripping the query/fragment from path:
I think there might be an issue for this already? We found this with a benchmark test.

``` 
bool PathMatcher::match(const absl::string_view path) const {
  return matcher_.match(Http::PathUtil::removeQueryAndFragment(path));
}
```

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

**Comment by daixiang0** [comment]
I think `matchRoute` is for HeaderMatcher and QueryParameterMatcher, not PathMatcher, they match different fields. The path may include query params but no header.

Could you share the benchmark result here?

**Comment by stevenzzzz** [comment]
you are right they match different fields, but I am not sure why that's relevant in this bug. 

The path_matcher actually only matches on <path>, not including query parameters or fragments.

The benchmark we did was internal, basically, caching the result of removeQueryAndFragment would end up with a performance improvement of nearly 5x difference.  

@AdiJohn could probably provide more details on the benchmark.  He actually filed an issue about the "PathMatcher::match" : http://go/envoy-upstream/issues/20352

**Comment by stevenzzzz** [comment]
@adisuissa do you think this worth a nostale tag?

**Comment by sschepens** [comment]
@stevenzzzz sure, I can do that. The only thing i'm a bit weary about query/parameters is that nowadays if someone isn't matching, we're not even parsing them, tough we can probably attach it the first time it's needed, does that seem right?

**Comment by stevenzzzz** [comment]
attaching on request sounds good to me. Tho I am wondering if this worth the extra complexity. 
Path matching is like 99% of routes we have seens, where by default the path_matcher strips the query parameters (if any) by default anyway.

**Comment by bryanwux** [comment]
Any updates on this? If it is really a 5x performance improvement, I think it's worth to do it. What about caching `sanitized_path` in somewhere else? @stevenzzzz 

https://github.com/envoyproxy/envoy/blob/4518b4920bd049d0e9f3cb5e519baed5fa2adf4c/source/common/router/config_impl.cc#L1567-L1568

**Comment by stevenzzzz** [comment]
Where exactly are you thinking?

I was hoping we can attach it to streamInfo which is 1:1, we could also make it part of the ActiveStream.

**Comment by bryanwux** [comment]
Sorry for the late reply. I agree with you that attach it to streamInfo will be a good idea. I will start implement the solution.
